source/common/http/alternate_protocols_cache_impl.cc

```diff
@@ -0,0 +1,51 @@
+#include "common/http/alternate_protocols_cache_impl.h"
+
+namespace Envoy {
+namespace Http {
+
+AlternateProtocolsCacheImpl::AlternateProtocolsCacheImpl(ThreadLocal::Instance& tls,
+                                                         TimeSource& time_source)
+    : time_source_(time_source), slot_(tls) {
+  slot_.set([](Event::Dispatcher& /*dispatcher*/) { return std::make_shared<State>(); });
+}
+
+AlternateProtocolsCacheImpl::~AlternateProtocolsCacheImpl() {}
+
+void AlternateProtocolsCacheImpl::setAlternatives(const Origin& origin,
+                                                  const std::vector<AlternateProtocol>& protocols,
+                                                  const MonotonicTime& expiration) {
+  // TODO(RyanTheOptimist): Propogate state changes across threads.
+  State& state = *(slot_.get());
+  Entry& entry = state.protocols_[origin];
+  if (entry.protocols_ != protocols) {
```

Comment:
Why the check before set? Is setting a lot more expensive than checking here for some reason? If yes, add a comment explaining why?

Reply:
Alyssa requested this check when this code was added (annoyingly, git isn't noticing that this file is basically a rename of the old alternate_protocols_cache.cc with a few changes. *sigh*). I'm happy to do whichever approach folks prefer.

Reply:
My .02 would be to have a comment explaining why the check-before-set pattern is being used, because my intuition is that it's an anti-pattern unless there is a reason for it. @alyssawilk is this because Envoy's thread-local storage is much cheaper to read than write?

Reply:
When you say "constant update", do you mean something else than "writing to RAM"? Does writing to this memory trigger another update mechanism that I'm not aware of?

Reply:
So naively we'd update the alt-svc cache on every response, since the ttl would be more fresh, and I didn't want to update all the strings on every response, just to update TTL.  As you say nothing as expensive as crypto but when you're a server handling tens of thousands of qps, it does make a per difference avoiding tens of thousands of string copies per second.